src/features/animation.ts

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {property} from 'lit-element';
+
+import ModelViewerElementBase, {$needsRender, $scene, $tick, $updateSource} from '../model-viewer-base.js';
+import {Constructor} from '../utils.js';
+
+const MILLISECONDS_PER_SECOND = 1000.0
+
+const $updateAnimation = Symbol('updateAnimation');
+const $updateModelLoadsPromise = Symbol('updateModelLoadsPromise');
+const $modelLoads = Symbol('modelLoads');
+const $rejectModelLoads = Symbol('rejectModelLoads');
+
+export const AnimationMixin =
+    (ModelViewerElement: Constructor<ModelViewerElementBase>):
+        Constructor<ModelViewerElementBase> => {
+          class AnimationModelViewerElement extends ModelViewerElement {
+            @property({type: Boolean}) animated: boolean = false;
+            @property({type: Boolean, attribute: 'pause-animation'})
+            pauseAnimation: boolean = false;
+            @property({type: String, attribute: 'animation-name'})
+            animationName: string|null = null;
+            @property({type: Number, attribute: 'animation-crossfade-duration'})
+            animationCrossfadeDuration: number = 300;
+
+            protected[$rejectModelLoads]: (...args: Array<any>) => void;
+            protected[$modelLoads]: Promise<void>;
+
+            constructor() {
+              super();
+              this[$updateModelLoadsPromise]();
+            }
+
+            /**
+             * Returns an array
+             */
+            get availableAnimations(): Array<string> {
+              if (this.loaded) {
+                return (this as any)[$scene].model.animationNames;
+              }
+
+              return [];
+            }
+
+            /**
+             * Creates a promise that resolves when a model is loaded. The
+             * promise resolves immediately if there is currently a loaded
+             * model. The promise rejects if the currently loaded model is
+             * changed to something new due to the src attribute changing.
+             */
+            [$updateModelLoadsPromise]() {
```

Comment:
Environments has a similar need/way of doing this -- wonder if there's a way to generalize this across all features?

Reply:
After offline discussion, I'm looking into whether this is necessary or not.

Reply:
I reduced the complexity of this implementation substantially, and I _think_ it should be fine. Thanks for calling this out!